js/web/lib/wasm/jsep/webgpu/ops/matmul.ts

```diff
@@ -27,36 +27,45 @@ export const createNaiveMatmulProgramInfo =
       const outerDims = reshapedOutputShape ? reshapedOutputShape.slice(0, -2) : outputShape.slice(0, -2);
       const batchSize = ShapeUtil.size(outerDims);
       const outputShapeInShader = [batchSize, M, N];
+
+      const batchDims = internalVariable('batch_dims', inputs[0].dataType, outerDims.length);
+      const a = inputVariable('a', inputs[0].dataType, aShape.length, aComponents);
+      const b = inputVariable('b', inputs[1].dataType, bShape.length, components);
+      const output = outputVariable('output', inputs[0].dataType, outputShapeInShader.length, components);
+      const {activationFunction, applyActivation} = getActivationSnippet(activationAttributes, output.type.value);
+      const inputVariables = [a, b];
+      let processBias = '';
+      if (hasBias) {
+        const biasComponents = isChannelsLast ? components : 1;
+        inputVariables.push(inputVariable('bias', inputs[2].dataType, inputs[2].dims.length, biasComponents));
+        processBias = `${
+            isChannelsLast ? `value += bias[col / ${biasComponents}];` :
+                             `value += ${output.type.value}(bias[row + i]);`}`;
+      }
+
+      const outerDimsA = aShape.slice(0, -2);
+      const outerDimsB = bShape.slice(0, -2);
+      const broadCastADims = getBroadcastDims(outerDimsA, outerDims);
+      const broadCastBDims = getBroadcastDims(outerDimsB, outerDims);
```

Comment:
Please don't move above code out of `getShaderSource`. Otherwise, they will be executed every time instead of once if the shader key not changed.

Reply:
Done